### PR TITLE
Implemented delete vector function based on the ID

### PR DESCRIPTION
**Pull request by erramuzpe**
This is a PR trying to solve #42 and #27 . I've only done the one for Memory storage, since I don't know much about Redis.

### CODE REVIEW

**Comment by amorgun** [comment]
I think you should add some tests.

**Comment by erramuzpe** [comment]
@amorgun I did the tests, could you take a look? thanks!

File: nearpy/tests/experiments_tests.py [hunk 1]
```diff
@@ -199,5 +199,69 @@ def test_random_binary_projections(self):
         print('\nRecall RBP: %f, Precision RBP: %f, SearchTime RBP: %f\n' % \
             (recall1, precision1, searchtime1))
 
+    def test_delete_vector(self):
+        # test case for 1 hash
+        dim = 5
+        rbp = RandomBinaryProjections('rbp', 10)
```
Comment (amorgun) [inline nearpy/tests/experiments_tests.py]: Try [UniBucket hash](https://github.com/pixelogik/NearPy/blob/master/nearpy/hashes/unibucket.py#L28), it's less random.

File: nearpy/tests/experiments_tests.py [hunk 2]
```diff
@@ -199,5 +199,69 @@ def test_random_binary_projections(self):
...
+        engine = Engine(dim, lshashes=[rbp])
+        # Index 20 random vectors (set their data to a unique string)
+        for index in range(20):
+            v = numpy.random.randn(dim)
+            engine.store_vector(v, index)
+        # Create random query vector
+        keys = set()
+        for lshash in engine.lshashes:
+            for value in engine.storage.buckets[lshash.hash_name].iteritems():
```
Comment (amorgun) [inline nearpy/tests/experiments_tests.py]: Python 3 does not support iteritems.

File: nearpy/storage/storage_memory.py [hunk 3]
```diff
@@ -49,6 +49,13 @@ def store_vector(self, hash_name, bucket_key, v, data):
             self.buckets[hash_name][bucket_key] = []
         self.buckets[hash_name][bucket_key].append((v, data))
 
+    def delete_vector(self, hash_name, data):
+        """
+        Deletes vector and JSON-serializable data in bucket with specified key.
+        """
+        for bucket_key in self.buckets[hash_name]:
+            self.buckets[hash_name][bucket_key] = [(v, id) for v, id in self.buckets[hash_name][bucket_key] if id != data]
```
Comment (amorgun) [inline nearpy/storage/storage_memory.py]: Do you intentionally remove all elements with the given data? Generally, data can be non-unique.

File: nearpy/tests/experiments_tests.py (hunk 2 above)
Comment (erramuzpe) [inline nearpy/tests/experiments_tests.py]: with just .items() will be enough?

File: nearpy/storage/storage_memory.py (hunk 3 above)
Comment (erramuzpe) [inline nearpy/storage/storage_memory.py]: As far as I understood, _data_ was an unique identifier for the vector _v_  ... 

Please, could you put me an example of non-uniqueness?

File: nearpy/tests/experiments_tests.py [hunk 4]
```diff
@@ -199,5 +199,69 @@ def test_random_binary_projections(self):
         print('\nRecall RBP: %f, Precision RBP: %f, SearchTime RBP: %f\n' % \
             (recall1, precision1, searchtime1))
 
+    def test_delete_vector(self):
```
Comment (amorgun) [inline nearpy/tests/experiments_tests.py]: Why is it in this file? I suggest placing it in the `storage_tests` or `engine_tests`.

File: nearpy/storage/storage_memory.py (hunk 3 above)
Comment (erramuzpe) [inline nearpy/storage/storage_memory.py]: Anyway, it is possible to add a (v, id) pair as input to identify a set and delete just the one matching.

Comment (amorgun) [inline nearpy/storage/storage_memory.py]: Example:
```python
import numpy
from nearpy import Engine
from nearpy.hashes import RandomBinaryProjections

dimension = 2
rbp = RandomBinaryProjections('rbp', 10)
engine = Engine(dimension, lshashes=[rbp])
v = numpy.random.randn(dimension)
for index in range(5):
    engine.store_vector(v, index)
N = engine.neighbours(v)
assert len(N) == 5
print(N)
```

Comment (amorgun) [inline nearpy/storage/storage_memory.py]: I fixed my example, but now I see that data is supposed to be unique.

File: nearpy/tests/engine_tests.py [hunk 5]
```diff
@@ -73,5 +73,69 @@ def test_retrieval_sparse(self):
...
+        # test case for 1 hash
+        dim = 5
+        unibucket = UniBucket('testHash')
+        engine = Engine(dim, lshashes=[unibucket])
+        # Index 20 random vectors (set their data to a unique string)
+        for index in range(20):
+            v = numpy.random.randn(dim)
+            engine.store_vector(v, index)
+        # Create random query vector
+        keys = set()
```
Comment (amorgun) [inline nearpy/tests/engine_tests.py]: You can just write
```python
bucket = engine.storage.buckets['testHash']['testHash']
keys = {i[1] for i in bucket}
```
I also suggest moving `'testHash'` into separate variable.

File: nearpy/tests/engine_tests.py [hunk 6]
```diff
@@ -73,5 +73,69 @@ def test_retrieval_sparse(self):
...
+        self.assertEqual(len(keys), 20)
+
+        engine.delete_vector(15) #delete the vector with key = 15
+
+        new_keys = set()
+        for lshash in engine.lshashes:
+            for value in engine.storage.buckets[lshash.hash_name].items():
+                [new_keys.add(id) for v, id in value[1]]
+
+        self.assertTrue(len(keys) > len(new_keys))
```
Comment (amorgun) [inline nearpy/tests/engine_tests.py]: This `assert` is redundant because of the next one.

File: nearpy/tests/engine_tests.py [hunk 7]
```diff
@@ -73,5 +73,69 @@ def test_retrieval_sparse(self):
...
+        self.assertEqual(len(new_keys), 19) #new keys has 19 elements instead of 20
+
+        self.assertTrue(15 in keys)
+        self.assertFalse(15 in new_keys) # the key 15 is the one missing
```
Comment (amorgun) [inline nearpy/tests/engine_tests.py]: Use `assertIn` and `assertNotIn`

File: nearpy/tests/engine_tests.py (hunk 5 above)
Comment (erramuzpe) [inline nearpy/tests/engine_tests.py]: thanks for the tip! changed

File: nearpy/tests/engine_tests.py [hunk 8]
```diff
@@ -73,5 +73,69 @@ def test_retrieval_sparse(self):
...
+
+    def test_delete_vector2(self):
```
Comment (amorgun) [inline nearpy/tests/engine_tests.py]: This name is not very descriptive. Move your comment from the next line into the test name.

File: nearpy/tests/engine_tests.py [hunk 9]
```diff
@@ -73,5 +73,68 @@ def test_retrieval_sparse(self):
...
+        # test case for several hashes
+        dim = 5
+        hashes = []
+        for k in xrange(10): # 10 hashes
```
Comment (amorgun) [inline nearpy/tests/engine_tests.py]: No `xrange` in `Python 3`

**Comment by pixelogik** [comment]
I am totally out guys. Please tell me when it makes sense to merge this. Thanks!

File: nearpy/tests/engine_tests.py (hunk 9 above)
Comment (erramuzpe) [inline nearpy/tests/engine_tests.py]: Aw, sorry for that. I just made a bunch of pushes... :sob:  Ok, I will wait for some minutes. thanks for your comments anyway! :+1:

File: nearpy/tests/engine_tests.py [hunk 10]
```diff
@@ -73,5 +73,58 @@ def test_retrieval_sparse(self):
...
+            v = numpy.random.randn(dim)
+            engine.store_vector(v, index)
+        # Create random query vector
+        keys = set()
+        bucket = engine.storage.buckets[name_hash][name_hash]
+        keys = {i[1] for i in bucket}
+
+        engine.delete_vector(15) #delete the vector with key = 15
+
+        new_keys = set()
```
Comment (amorgun) [inline nearpy/tests/engine_tests.py]: You don't need to declare variables like this and `keys` before assignment.

File: nearpy/tests/engine_tests.py [hunk 11]
```diff
@@ -73,5 +73,58 @@ def test_retrieval_sparse(self):
...
+        bucket = engine.storage.buckets[name_hash][name_hash]
```
Comment (amorgun) [inline nearpy/tests/engine_tests.py]: You use the same code twice (and two more times in the next test). You should probably think about moving it into method, like `get_keys(engine, hash_name)`

File: nearpy/tests/engine_tests.py [hunk 12]
```diff
@@ -73,5 +73,58 @@ def test_retrieval_sparse(self):
...
+        engine = Engine(dim, lshashes=hashes)
+        # Index 20 random vectors (set their data to a unique string)
+        for index in range(20):
+            v = numpy.random.randn(dim)
+            engine.store_vector(v, index)
+        # Create random query vector
+        keys = set()
+        for lshash in engine.lshashes:
+            bucket = engine.storage.buckets[lshash.hash_name][lshash.hash_name]
+            keys = {i[1] for i in bucket}
```
Comment (amorgun) [inline nearpy/tests/engine_tests.py]: You rewrite `keys` here. I think it shoud be `keys.update(keys_from_bucket)` instead.

File: nearpy/tests/engine_tests.py [hunk 13]
```diff
@@ -73,5 +73,58 @@ def test_retrieval_sparse(self):
             self.assertEqual(y_data, x_data)
             self.assertAlmostEqual(y_distance, 0.0, delta=delta)
 
+    def test_delete_vector_single_hash(self):
+        dim = 5
+        name_hash = 'testHash'
+        unibucket = UniBucket(name_hash)
```
Comment (amorgun) [inline nearpy/tests/engine_tests.py]: Seriously, try running your tests. You forgot to import `UniBucket`.

File: nearpy/tests/engine_tests.py [hunk 14]
```diff
@@ -73,5 +73,58 @@ def test_retrieval_sparse(self):
...
+        engine = Engine(dim, lshashes=[unibucket])
+        # Index 20 random vectors (set their data to a unique string)
+        for index in range(20):
+            v = numpy.random.randn(dim)
```
Comment (amorgun) [inline nearpy/tests/engine_tests.py]: I am personally against randomness in tests and prefer constant vectors (like `np.ones(dim) * index`), but if you like random vectors you must [set random state](https://github.com/pixelogik/NearPy/blob/master/nearpy/tests/experiments_tests.py#L37-L38).

File: nearpy/tests/engine_tests.py (hunk 13 above)
Comment (erramuzpe) [inline nearpy/tests/engine_tests.py]: Sorry, :neutral_face: first time writing tests...

File: nearpy/tests/engine_tests.py (hunk 14 above)
Comment (erramuzpe) [inline nearpy/tests/engine_tests.py]: Yep, I understand, but random vectors here don't really matter for the test I guess... Anyway, thanks for the info! :)

**Comment by erramuzpe** [comment]
I will clean this up!

File: nearpy/utils/utils.py [hunk 15]
```diff
@@ -89,3 +89,9 @@ def want_string(arg, encoding='utf-8'):
     else:
         rv = arg
     return rv
+
+
+def get_keys(engine):
```
Comment (amorgun) [inline nearpy/utils/utils.py]: This function in its current state makes sense only in tests and only for engine with memory storage and unibucket hashes. I initially suggested moving it into method of the test suite (methods with names that are not started with "test_" are not treated as tests and can be used as helpers). Ideally you can make your own suite with custom setup and helpers, but I'm fine with leaving all in the `engine_tests`.

**Pull request closed without merging**